README.md
# goc25519sm

----

A Go package for validated and consensus-safe Curve25519 scalar multiplication.

----

## Availability

- [GitHub](https://github.com/johnsonjh/goc25519sm)
- [GitLab (HQ)](https://gitlab.com/johnsonjh/goc25519sm)
- [GitLab (trnsz)](https://gitlab.trnsz.com/johnsonjh/goc25519sm)
- [SourceHut](https://sr.ht/~trn/goc25519sm)
- [NotABug](https://notabug.org/trn/goc25519sm)

----

## License

- [BSD 3-Clause License (Revised)](https://tldrlegal.com/license/bsd-3-clause-license-(revised))

----


CHANGELOG.md
2021-03-06
==========

  * Add CHANGELOG.md: Autogenerated by `git-changelog`
  * Update README.md: Add self-hosted GitLab mirror.
  * Update dependencies.
  * Update README.md: Add Availability section w/links
  * multi: Multiple cosmetic cleanups; update deps.
  * Update dependencies, fixup import paths, etc.
  * Another import path fixup
  * multi: use github for imports, etc.
  * README: Remove (unused) mailing list information.

1.6.4 / 2021-02-15
==================

  * Bump deps, update leaktestfe, CI/CD configuration

1.6.3 / 2021-02-14
==================

  * Merge pull request #43 from gridfinity/deepsource-transform-daf3ba43
  * Update L2 additional dependencies
  * Format code with black, gofmt, yapf, autopep8, isort, standardrb,
    standardjs, prettier and rubocop
  * Update license, copyright, docs, and issue info.

1.6.2 / 2021-01-24
==================

  * Remove deprecreated dependency, update leaktestfe.
  * Update tools dependencies
  * Update for 1.16+ compiler compatability, new ltfe
  * Merge pull request #42 from gridfinity/rm-license-block-readme
  * Update README.md
  * Merge pull request #41 from gridfinity/dsr
  * Update README.md
  * CI/CD: Simplify build
  * Happy New Year
  * Depreciate --ignore-cr-at-eol
  * Update dependenices
  * Merge pull request #40 from
    gridfinity/restyled/deepsource-transform-b7b853ad
  * Merge pull request #39 from gridfinity/deepsource-transform-b7b853ad
  * Restyled by prettier-markdown
  * Restyled by prettier-json
  * Restyled by jq
  * Format code with black, gofmt, yapf, autopep8, isort, standardrb,
    standardjs, prettier and rubocop
  * Update .deepsource.toml
  * Merge pull request #38 from gridfinity/small-fossa
  * Update FOSSA license status
  * Merge pull request #37 from gridfinity/COVERALLS
  * Coveralls!
  * Merge pull request #36 from gridfinity/update-covralls-link
  * README: Upate Coveralls Badge.
  * Merge pull request #35 from gridfinity/fossa-badge-1
  * Update upper FOSSA badge link
  * Merge pull request #34 from gridfinity/godoc-link
  * Fix GoDoc link
  * Merge pull request #33 from gridfinity/coveralls-new-badge
  * Correct Coveralls.io link
  * Merge pull request #32 from gridfinity/remove-tag-badge
  * Removed Tag badge.
  * Merge pull request #31 from gridfinity/layout-spacing-markdown
  * Tweak badge layout
  * Merge pull request #30 from gridfinity/badge-urls
  * Additional badge update URLs
  * Merge pull request #29 from gridfinity/codebeat-badge
  * Update CodeBeat badge
  * Merge pull request #28 from gridfinity/badge-link-updates
  * Badge link updates
  * Merge pull request #27 from gridfinity/restyled/lgtm-badge
  * Merge pull request #26 from gridfinity/lgtm-badge
  * Restyled by prettier-markdown
  * Update LGTM Alert Badge
  * Update leaktestfe dependency
  * Mergify: configuration update
  * Update Dependabot config file
  * Rename testutil -> leaktestfe, move to new API
  * Switch over to new Leakplug API in leaktestfe
  * More adjustments for new import paths
  * Update documentation for new important path.
  * Update .gitignore
  * Fix embedded testutil to -> leaktestfe package
  * Updated tested/trusted dependency set.
  * CI: Use full path to rpmconf
  * CI: For Linux, do a groupupdate for base
  * CI: Add ccache, pv, valgrind, graphviz for Linux
  * cov: Warn cov\_report.sh is not POSIX compliant
  * Remove unnecessary comment
  * Update licensing, compile-inlicense, add test.
  * Update golegal import
  * cov: Reformat to appease shellcheck all + shfmt
  * CI: On Windows, test -a w/race requires GCC in path.
  * CI: Unbreak macOS
  * Fix for Windows paths
  * More GitLab CI/CD Go integration on OS X [2/2]
  * More GitLab CI/CD Go integration on OS X
  * GitLab CI/CD fixes for new runner configs
  * CI: Unbreak concurrent macOS builds
  * Update dependencies, adjust for new CI/CD tags
  * Merge pull request #23 from johnsonjh/lic-2-fossa
  * README: review -> see, all -> full
  * Merge pull request #22 from johnsonjh/fossa-relocate
  * README: Relocate secondary [small] FOSSA badge
  * README: Deepsource status
  * SECURITY: Use explicit mailto to appease linter
  * testutil: Comment to document leak wrapper funcs
  * SECURITY: Formatting
  * README: Link to SECURITY: Security Policy document
  * Merge pull request #20 from johnsonjh/shiftleft-action-config-1607967495
  * Update actions/checkout action to v2
  * adding ShiftLeft action workflow config
  * SECURITY: Appease markdownlint
  * Remove unnecessary comparison with bool
  * Merge pull request #19 from johnsonjh/security-policy
  * Merge pull request #17 from johnsonjh/codeql-analysis
  * Merge pull request #16 from johnsonjh/codacy-analysis
  * Merge pull request #15 from johnsonjh/shiftleft-scan
  * Create SECURITY.md
  * Create shiftleft-analysis.yml
  * Create codacy-analysis.yml
  * Create codeql-analysis.yml
  * README: Remove GitLab CI status badge, appese lint
  * README: Add alt-text for CI/CD
  * README: Update GitLab CI/CD pipeline status
  * README: Appease markdownlint
  * README: Add missing punctuation
  * goc25519sm: Update internal license text
  * CI: Minor update for comment consistency
  * CI: Unbreak legacy git
  * README: Update coveralls.io status badge
  * goc25519sm\_generic: 'and' -> 'but'
  * goc25519sm: Update GoDocs text
  * README: Clarify licensing terms and conditions
  * LICENSE: Additional attributions
  * CI: Remove -buildid= for current test builds
  * CI: Extra git config, explicit Go build/test order
  * multi: Many improvements throughout the package
  * CI: git -P -> --no-pager for backwards compat
  * CI: Fixup Linux build, add Git clean block
  * CI: Update Linux development tools installation
  * CI: Use go list to make targets + PowerShell fixes.
  * CI: Specify tags, extra environment information
  * README: Adjust GitLab CI Pipeline status badge
  * README: Markdown adjustments
  * README: GitLab pipeline status
  * go: Update dependencies
  * README: Fix FOSSA table

goc25519sm_amd64.go
// Copyright © 2021 Jeffrey H. Johnson <trnsz@pobox.com>.
// Copyright © 2021 Gridfinity, LLC.
// Copyright © 2021 Filippo Valsorda.
// Copyright © 2012 The Go Authors.
//
// All rights reserved.
//
// Use of this source code is governed by the BSD-style
// license that can be found in the LICENSE file.
//
//go:build amd64 && gc && !purego
// +build amd64,gc,!purego

package goc25519sm

import (
	"fmt"
)

// These functions are implemented in the '.s' files
// Naming is analogous to the SUPERCOP implementation

//go:noescape
func cswap(
	inout *[5]uint64,
	v uint64,
)

//go:noescape
func ladderstep(
	inout *[5][5]uint64,
)

//go:noescape
func freeze(
	inout *[5]uint64,
)

//go:noescape
func mul(
	dest,
	a,
	b *[5]uint64,
)

//go:noescape
func square(
	out,
	in *[5]uint64,
)

// mladder implements a Montgomery ladder
// to calculate ( ( 'xr'/'zr' ) *= 's' )
func mladder(
	xr,
	zr *[5]uint64,
	s *[X25519Size]byte,
) {
	var work [5][5]uint64
	work[0] = *xr
	setint(
		&work[1],
		1,
	)
	setint(
		&work[2],
		0,
	)
	work[3] = *xr
	setint(
		&work[4],
		1,
	)
	j := uint(6)
	var prevbit byte
	for i := 31; i >= 0; i-- {
		for j < 8 {
			bit := ((*s)[i] >> j) & 1
			swap := bit ^ prevbit
			prevbit = bit
			cswap(
				&work[1],
				uint64(swap),
			)
			ladderstep(
				&work,
			)
			j--
		}
		j = 7
	}
	*xr = work[1]
	*zr = work[2]
}

func oldScalarMult(
	dst,
	scalar,
	base *[X25519Size]byte,
) error {
	var e [X25519Size]byte
	var err error
	// Dubious to perform clamping at this stage,
	// but the behavior matches that of libsodium
	copy(
		e[:],
		(*scalar)[:],
	)
	e[0] &= 248
	e[31] &= 127
	e[31] |= 64
	var t,
		z [5]uint64
	unpack(
		&t,
		base,
	)
	mladder(
		&t,
		&z,
		&e,
	)
	invert(
		&z,
		&z,
	)
	mul(
		&t,
		&t,
		&z,
	)
	pack(
		dst,
		&t,
	)
	err = oldScalarMultVerify(
		dst,
		scalar,
		base,
	)
	if err != nil {
		return fmt.Errorf(
			"\ngoc25519sm.oldScalarMult.OldScalarMult_amd64.oldScalarMultVerify FAILURE:\n	dst=%v\n	scalar=%v\n	base=%v\n	%v",
			*dst,
			*scalar,
			*base,
			err,
		)
	}
	return nil
}

func setint(
	r *[5]uint64,
	v uint64,
) {
	r[0] = v
	r[1] = 0
	r[2] = 0
	r[3] = 0
	r[4] = 0
}

// unpack sets 'r' = 'x' where 'r' consists of
// five 51-bit limbs (in little-endian order)
func unpack(
	r *[5]uint64,
	x *[X25519Size]byte,
) {
	r[0] = uint64(x[0]) |
		uint64(x[1])<<8 |
		uint64(x[2])<<16 |
		uint64(x[3])<<24 |
		uint64(x[4])<<32 |
		uint64(x[5])<<40 |
		uint64(x[6]&7)<<48

	r[1] = uint64(x[6])>>3 |
		uint64(x[7])<<5 |
		uint64(x[8])<<13 |
		uint64(x[9])<<21 |
		uint64(x[10])<<29 |
		uint64(x[11])<<37 |
		uint64(x[12]&63)<<45

	r[2] = uint64(x[12])>>6 |
		uint64(x[13])<<2 |
		uint64(x[14])<<10 |
		uint64(x[15])<<18 |
		uint64(x[16])<<26 |
		uint64(x[17])<<34 |
		uint64(x[18])<<42 |
		uint64(x[19]&1)<<50

	r[3] = uint64(x[19])>>1 |
		uint64(x[20])<<7 |
		uint64(x[21])<<15 |
		uint64(x[22])<<23 |
		uint64(x[23])<<31 |
		uint64(x[24])<<39 |
		uint64(x[25]&15)<<47

	r[4] = uint64(x[25])>>4 |
		uint64(x[26])<<4 |
		uint64(x[27])<<12 |
		uint64(x[28])<<20 |
		uint64(x[29])<<28 |
		uint64(x[30])<<36 |
		uint64(x[31]&127)<<44
}

// pack sets 'out' = 'x' where 'out' is the standard
// little-endian form of the five 51-bit limbs in 'x'
func pack(
	out *[X25519Size]byte,
	x *[5]uint64,
) {
	t := *x
	freeze(
		&t,
	)
	out[0] = byte(t[0])
	out[1] = byte(t[0] >> 8)
	out[2] = byte(t[0] >> 16)
	out[3] = byte(t[0] >> 24)
	out[4] = byte(t[0] >> 32)
	out[5] = byte(t[0] >> 40)
	out[6] = byte(t[0] >> 48)
	out[6] ^= byte(t[1]<<3) & 0xf8
	out[7] = byte(t[1] >> 5)
	out[8] = byte(t[1] >> 13)
	out[9] = byte(t[1] >> 21)
	out[10] = byte(t[1] >> 29)
	out[11] = byte(t[1] >> 37)
	out[12] = byte(t[1] >> 45)
	out[12] ^= byte(t[2]<<6) & 0xc0
	out[13] = byte(t[2] >> 2)
	out[14] = byte(t[2] >> 10)
	out[15] = byte(t[2] >> 18)
	out[16] = byte(t[2] >> 26)
	out[17] = byte(t[2] >> 34)
	out[18] = byte(t[2] >> 42)
	out[19] = byte(t[2] >> 50)
	out[19] ^= byte(t[3]<<1) & 0xfe
	out[20] = byte(t[3] >> 7)
	out[21] = byte(t[3] >> 15)
	out[22] = byte(t[3] >> 23)
	out[23] = byte(t[3] >> 31)
	out[24] = byte(t[3] >> 39)
	out[25] = byte(t[3] >> 47)
	out[25] ^= byte(t[4]<<4) & 0xf0
	out[26] = byte(t[4] >> 4)
	out[27] = byte(t[4] >> 12)
	out[28] = byte(t[4] >> 20)
	out[29] = byte(t[4] >> 28)
	out[30] = byte(t[4] >> 36)
	out[31] = byte(t[4] >> 44)
}

// invert calculates 'r' = (('x'^-1) mod 'p')
// using Fermat's little theorem
func invert(
	r,
	x *[5]uint64,
) {
	var z2,
		z9,
		z11,
		z2_5_0,
		z2_10_0,
		z2_20_0,
		z2_50_0,
		z2_100_0,
		t [5]uint64
	square(
		&z2,
		x,
	) // 2
	square(
		&t,
		&z2,
	) // 4
	square(
		&t,
		&t,
	) // 8
	mul(
		&z9,
		&t,
		x,
	) // 9
	mul(
		&z11,
		&z9,
		&z2,
	) // 11
	square(
		&t,
		&z11,
	) // 22
	mul(
		&z2_5_0,
		&t,
		&z9,
	) // 2^5 - 2^0 = 31
	square(
		&t,
		&z2_5_0,
	) // 2^6 - 2^1
	for i := 1; i < 5; i++ {
		square(
			&t,
			&t,
		)
	} // 2^20 - 2^10
	mul(
		&z2_10_0,
		&t,
		&z2_5_0,
	) // 2^10 - 2^0
	square(
		&t,
		&z2_10_0,
	) // 2^11 - 2^1
	for i := 1; i < 10; i++ {
		square(
			&t,
			&t,
		)
	} // 2^20 - 2^10
	mul(
		&z2_20_0,
		&t,
		&z2_10_0,
	) // 2^20 - 2^0
	square(
		&t,
		&z2_20_0,
	) // 2^21 - 2^1
	for i := 1; i < 20; i++ {
		square(
			&t,
			&t,
		)
	} // 2^40 - 2^20
	mul(
		&t,
		&t,
		&z2_20_0,
	) // 2^40 - 2^0
	square(
		&t,
		&t,
	) // 2^41 - 2^1
	for i := 1; i < 10; i++ {
		square(
			&t,
			&t,
		)
	} // 2^50  - 2^10
	mul(
		&z2_50_0,
		&t,
		&z2_10_0,
	) // 2^50  - 2^0
	square(
		&t,
		&z2_50_0,
	) // 2^51  - 2^1
	for i := 1; i < 50; i++ {
		square(
			&t,
			&t,
		)
	} // 2^100 - 2^50
	mul(
		&z2_100_0,
		&t,
		&z2_50_0,
	) // 2^100 - 2^0
	square(
		&t,
		&z2_100_0,
	) // 2^101 - 2^1
	for i := 1; i < 100; i++ {
		square(
			&t,
			&t,
		)
	} // 2^200 - 2^100
	mul(
		&t,
		&t,
		&z2_100_0,
	) // 2^200 - 2^0
	square(
		&t,
		&t,
	) // 2^201 - 2^1
	for i := 1; i < 50; i++ {
		square(
			&t,
			&t,
		)
	} // 2^250 - 2^50
	mul(
		&t,
		&t,
		&z2_50_0,
	) // 2^250 - 2^0
	square(
		&t,
		&t,
	) // 2^251 - 2^1
	square(
		&t,
		&t,
	) // 2^252 - 2^2
	square(
		&t,
		&t,
	) // 2^253 - 2^3
	square(
		&t,
		&t,
	) // 2^254 - 2^4
	square(
		&t,
		&t,
	) // 2^255 - 2^5
	mul(
		r,
		&t,
		&z11,
	) // 2^255 - 21
}
